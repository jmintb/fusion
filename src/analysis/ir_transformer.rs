use std::collections::{BTreeMap, BTreeSet, VecDeque};

use crate::ir::{Block, IrProgram, Variable, SSAID};
use crate::ast::nodes::AccessModes;

use crate::{control_flow_graph::ControlFlowGraph, ir::BlockId};
use anyhow::Result;
use tracing::debug;

use super::borrow_checker::VariableState;

#[derive(Clone, Debug)]
pub struct IrScope {
    pub blocks: BTreeMap<BlockId, Block>,
    pub control_flow_graph: ControlFlowGraph<BlockId>,
}

#[derive(Clone, Debug)]
pub struct IrInterpreter<Ctx> {
    scope: IrScope,
    control_flow_graph: ControlFlowGraph<BlockId>,
    block_states: BTreeMap<BlockId, Ctx>,
    pub ssa_variables: BTreeMap<SSAID, Variable>,
    pub access_modes: BTreeMap<SSAID, AccessModes>,
    context: Ctx,
    reverse_traversel: bool,
}

pub struct ReverseCycleAwareBlockIterator {
    qeueue: Vec<Vec<BlockId>>,
}

impl ReverseCycleAwareBlockIterator {
    fn new(control_flow_graph: ControlFlowGraph<BlockId>) -> Self {
        Self {
            qeueue: {
                let mut successors = control_flow_graph
                    .cycle_aware_successors(&control_flow_graph.entry_point)
                    .unwrap();
                successors.insert(0, vec![control_flow_graph.entry_point]);
                successors
            },
        }
    }
}

impl Iterator for ReverseCycleAwareBlockIterator {
    type Item = Vec<BlockId>;

    fn next(&mut self) -> Option<Self::Item> {
        self.qeueue.pop()
    }
}

pub struct IrBlockIterator {
    queue: VecDeque<BlockId>,
    visited_blocks: BTreeSet<BlockId>,
    control_flow_graph: ControlFlowGraph<BlockId>,
    visit_again: BTreeSet<BlockId>,
}

impl IrBlockIterator {
    fn new(control_flow_graph: ControlFlowGraph<BlockId>) -> Self {
        let mut queue = VecDeque::new();
        queue.push_back(control_flow_graph.entry_point);
        Self {
            control_flow_graph,
            queue,
            visited_blocks: BTreeSet::new(),
            visit_again: BTreeSet::new(),
        }
    }
}

impl Iterator for IrBlockIterator {
    type Item = BlockId;

    fn next(&mut self) -> Option<Self::Item> {
        let mut next = self.queue.pop_front()?;
        while self.visited_blocks.contains(&next) && !self.visit_again.contains(&next) {
            next = self.queue.pop_front()?;
        }

        // self.visited_blocks.remove(&next);

        debug!("next: {}, queue {:?}", next.0, self.queue);

        // TODO: cycles  seem to be handled , but we need a rigorous test and to avoid the extra visits currently happening.
        if !self.visited_blocks.contains(&next) {
            for predecessor in self.control_flow_graph.direct_predecessors(&next.clone()) {
                if !self.visited_blocks.contains(&predecessor) {
                    if self.control_flow_graph.dominates(next, predecessor) {
                        let predecessors_preds = self
                            .control_flow_graph
                            .predecessors(&predecessor)
                            .unwrap()
                            .into_iter()
                            .flatten()
                            .collect::<Vec<Self::Item>>();

                        debug!(
                            "pred preds: {:?} \n  {} {}",
                            predecessors_preds, predecessor, next
                        );

                        let Some(loop_start_pos) =
                            predecessors_preds.iter().position(|pred| *pred == next)
                        else {
                            break;
                        };

                        let mut loop_blocks = predecessors_preds;
                        loop_blocks.split_off(loop_start_pos);
                        loop_blocks.reverse();
                        loop_blocks.push(predecessor);

                        debug!("cycle blocks: {:?}", loop_blocks);
                        assert!(!loop_blocks.contains(&next));

                        let mut loop_successors: Vec<BlockId> = self
                            .control_flow_graph
                            .successors(&loop_blocks[0])
                            .unwrap()
                            .into_iter()
                            .flatten()
                            .collect();

                        let cycle_completetion_position = loop_successors
                            .iter()
                            .position(|pred| *pred == next)
                            .unwrap();

                        loop_successors.split_off(cycle_completetion_position);

                        // loop_successors
                        //     .clone()
                        //     .into_iter()
                        //     .for_each(|block| loop_blocks.push(block));

                        self.visited_blocks.insert(next);
                        self.visit_again.insert(next);
                        if loop_blocks.len() == 1 {
                            self.queue.push_front(loop_blocks[0]);
                            self.queue.push_front(next);
                            self.visit_again.insert(loop_blocks[0]);
                            next = loop_blocks[0];
                            continue;
                        }

                        let (loop_start, loop_blocks) = loop_blocks.split_first().unwrap();
                        debug!(
                            "loop successor blocks: {:?} {:?} {} {}",
                            loop_successors, loop_blocks, loop_start, predecessor
                        );
                        let mut loop_blocks = loop_blocks.to_vec().clone();

                        loop_blocks.reverse();

                        self.queue.push_front(*loop_start);
                        for loop_predecessor in loop_blocks.clone() {
                            self.queue.push_front(loop_predecessor);
                            self.visit_again.insert(loop_predecessor);
                        }

                        self.queue.push_front(next);
                        self.queue.push_front(*loop_start);
                        self.visit_again.insert(*loop_start);
                        for loop_predecessor in loop_blocks {
                            self.queue.push_front(loop_predecessor);
                        }
                        debug!("skipping {next} {loop_start}");

                        // TODO: remove predecssor from grandchildren, as we end up visiting it again.
                        // This logic tricky, as we need to ensure we revisit the dominated block after visting the dominator.
                        next = *loop_start;
                        break; // TODO: Continuing probably doesn't make sense here as we could have multiple unvisited predecessors I think.
                    } else {
                        // todo!(
                        //     "this should not be possible, but how do we handle it? ,{:#?} {} {}",
                        //     self.control_flow_graph,
                        //     predecessor.0,
                        //     next.0
                        // );
                    }
                }
            }
        }
        debug!("processing block: {}", next.0);

        for &grand_child in self
            .control_flow_graph
            .graph
            .get(&next)
            .unwrap_or(&Vec::new())
        {
            if !self.visited_blocks.contains(&grand_child) {
                self.queue.push_back(grand_child);
            }
        }

        self.visited_blocks.insert(next);

        // sleep(Duration::from_secs(1));

        Some(next)
    }
}

impl IntoIterator for ControlFlowGraph<BlockId> {
    type Item = BlockId;

    type IntoIter = IrBlockIterator;

    fn into_iter(self) -> Self::IntoIter {
        IrBlockIterator::new(self)
    }
}

impl ControlFlowGraph<BlockId> {
    fn into_reverse_cycle_aware_iterator(self) -> ReverseCycleAwareBlockIterator {
        ReverseCycleAwareBlockIterator::new(self)
    }
}

pub struct TransformContext {
    pub scope: IrScope,
    pub ssa_variables: BTreeMap<SSAID, Variable>,
    pub access_modes: BTreeMap<SSAID, AccessModes>,
    pub variable_states: BTreeMap<SSAID, VariableState>,
}

type TransformFn<Ctx> =
    dyn FnMut(usize, &mut TransformContext, &BlockId, &mut Ctx) -> Result<usize>;

impl<Ctx: Clone + Default> IrInterpreter<Ctx> {
    pub fn new(control_flow_graph: ControlFlowGraph<BlockId>, program: IrProgram) -> Self {
        let scope = IrScope {
            blocks: program.blocks,
            control_flow_graph: control_flow_graph.clone(),
        };
        Self {
            control_flow_graph,
            scope,
            ssa_variables: program.ssa_variables,
            access_modes: program.access_modes,
            block_states: BTreeMap::new(),
            context: Ctx::default(),
            reverse_traversel: false,
        }
    }

    pub fn new_reversed(control_flow_graph: ControlFlowGraph<BlockId>, program: IrProgram) -> Self {
        let scope = IrScope {
            blocks: program.blocks,
            control_flow_graph: control_flow_graph.clone(),
        };
        Self {
            control_flow_graph,
            scope,
            ssa_variables: program.ssa_variables,
            access_modes: program.access_modes,
            block_states: BTreeMap::new(),
            context: Ctx::default(),
            reverse_traversel: true,
        }
    }

    pub fn transform(mut self, transform_fn: &mut TransformFn<Ctx>) -> Result<Ctx> {
        let mut ctx = TransformContext {
            scope: self.scope.clone(),
            ssa_variables: self.ssa_variables.clone(),
            access_modes: self.access_modes.clone(),
            variable_states: BTreeMap::new(),
        };

        if self.reverse_traversel {
            for block in self
                .control_flow_graph
                .clone()
                .into_reverse_cycle_aware_iterator()
                .flatten()
            {
                self.transform_block(block, transform_fn, &mut ctx)?;
            }
        } else {
            // for block in self.control_flow_graph.clone() {
            //     debug!("block order {}", block);
            // }
            for block in self.control_flow_graph.clone() {
                self.transform_block(block, transform_fn, &mut ctx)?;
            }
        }

        Ok(self.context)
    }

    fn transform_block(
        &mut self,
        block_id: BlockId,
        transform_fn: &mut TransformFn<Ctx>,
        ctx: &mut TransformContext,
    ) -> Result<()> {
        let block = self.scope.blocks.get_mut(&block_id).unwrap();
        if block.instructions.is_empty() && self.reverse_traversel {
            return Ok(());
        }

        let mut instruction_counter = if self.reverse_traversel {
            block.instructions.len() - 1
        } else {
            0
        };

        loop {
            instruction_counter =
                transform_fn(instruction_counter, ctx, &block_id, &mut self.context)?;

            if (instruction_counter == 0 && !self.reverse_traversel)
                || (self.reverse_traversel && instruction_counter == block.instructions.len())
            {
                break;
            }
        }

        Ok(())
    }
}
