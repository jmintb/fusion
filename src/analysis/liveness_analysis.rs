use std::collections::{BTreeMap, BTreeSet};

use anyhow::{Result};

use tracing::debug;

use crate::{
    control_flow_graph::ControlFlowGraph,
    ir::{BlockId, FunctionId, Instruction, IrProgram, SSAID}, ast::identifiers::FunctionDeclarationID,
};

use super::ir_transformer::IrInterpreter;

#[derive(Clone, Debug, Copy)]
pub struct AbstractAddress {
    pub block_id: BlockId,
    pub inststruction: usize,
}

#[derive(Clone, Debug)]
pub struct AbstractAddressRange {
    pub start_address: AbstractAddress,
    pub end_addresses: Vec<AbstractAddress>,
}

impl AbstractAddressRange {
    fn insert_end(&mut self, address: AbstractAddress) {
        self.end_addresses.push(address);
    }

    fn new(start_address: AbstractAddress) -> Self {
        Self {
            start_address,
            end_addresses: vec![start_address],
        }
    }
}

#[derive(Clone, Debug, Default)]
pub struct VariableLiveness {
    pub variables: BTreeMap<SSAID, AbstractAddressRange>,
    pub variable_moved: BTreeMap<SSAID, BTreeSet<BlockId>>,
    pub loans: BTreeMap<SSAID, Vec<SSAID>>,
}

impl VariableLiveness {
    fn new() -> Self {
        Self {
            variables: BTreeMap::new(),
            loans: BTreeMap::new(),
            variable_moved: BTreeMap::new(),
        }
    }

    fn insert_variable_start(
        &mut self,
        variable_id: SSAID,
        address: AbstractAddress,
    ) -> Result<()> {
        self.variables
            .entry(variable_id)
            .and_modify(|entry| entry.start_address = address)
            .or_insert(AbstractAddressRange::new(address));

        Ok(())
    }

    fn insert_move(&mut self, variable_id: SSAID, block_id: BlockId) {
        self.variable_moved
            .entry(variable_id)
            .and_modify(|drop_ids| {
                drop_ids.insert(block_id);
            })
            .or_insert(BTreeSet::from_iter(vec![block_id].into_iter()));
    }

    fn insert_variable_end(
        &mut self,
        id: SSAID,
        address: AbstractAddress,
        moved: bool,
        control_flow_graph: &ControlFlowGraph<BlockId>,
    ) -> Result<()> {
        let _variables = self.variables.clone();
        let Some(entry) = self.variables.get_mut(&id) else {
            self.variables
                .insert(id, AbstractAddressRange::new(address));
            if moved {
                self.insert_move(id, address.block_id);
            }
            return Ok(());
        };

        if let Some(_end_in_current_block) = entry
            .end_addresses
            .iter()
            .position(|end_address| end_address.block_id == address.block_id)
        {
            return Ok(());
        } else {
            let entry_clone = entry.clone();
            if entry_clone
                .end_addresses
                .iter()
                .enumerate()
                .any(|(_, end_address)| {
                    control_flow_graph
                        .cycle_aware_successors(&address.block_id)
                        .unwrap()
                        .into_iter()
                        .flatten()
                        .collect::<Vec<BlockId>>()
                        .contains(&end_address.block_id)
                })
            {
                return Ok(());
            }
            entry.end_addresses.push(address);
            if moved {
                self.insert_move(id, address.block_id);
            }
        }

        Ok(())
    }

    // TODO: next use reverse traverself to calcuate liveness. If two blocks diverge keep end addresses.

    pub fn variabled_moved(&self, id: &SSAID, block_id: BlockId) -> bool {
        self.variable_moved
            .get(id)
            .map(|block_ids| block_ids.contains(&block_id))
            .unwrap_or(false)
    }
}

pub fn calculate_livenss(ir_program: &IrProgram) -> Result<BTreeMap<FunctionDeclarationID, VariableLiveness>> {
    debug!("calculating livenss for program: {ir_program}");

    let mut liveness_for_fn = BTreeMap::new();
    for (function_id, control_flow_graph) in ir_program.control_flow_graphs.clone() {
        debug!(
            "control flow graph: {}: {}\n predecessors: {:#?}",
            function_id.0,
            control_flow_graph,
            control_flow_graph
                .cycle_aware_successors(&control_flow_graph.entry_point)
                .unwrap()
        );
        let interpreter = IrInterpreter::<VariableLiveness>::new_reversed(
            ir_program
                .control_flow_graphs
                .get(&function_id)
                .cloned()
                .unwrap(),
            ir_program.clone(),
        );

        let livenss = interpreter.transform(
            &mut |instruction_counter, ctx, block_id, variable_livness| {
                debug!(
                    "liveness for {} at instruction {}",
                    block_id, instruction_counter
                );
                let block = ctx.scope.blocks.get(block_id).unwrap();
                let Some(instruction) = block.instructions.get(instruction_counter) else {
                    return Ok(block.instructions.len());
                };

                match instruction {
                    Instruction::Assign(to, from) => variable_livness.insert_variable_start(
                        *to,
                        AbstractAddress {
                            block_id: *block_id,
                            inststruction: instruction_counter,
                        },
                    )?,
                    Instruction::BorrowEnd(id) | Instruction::MutBorrowEnd(id) => {
                        variable_livness.insert_variable_end(
                            *id,
                            AbstractAddress {
                                block_id: *block_id,
                                inststruction: instruction_counter,
                            },
                            false,
                            &ctx.scope.control_flow_graph,
                        )?;
                    }
                    Instruction::Move(id) | Instruction::Drop(id) => {
                        variable_livness.insert_variable_end(
                            *id,
                            AbstractAddress {
                                block_id: *block_id,
                                inststruction: instruction_counter,
                            },
                            true,
                            &ctx.scope.control_flow_graph,
                        )?;
                    }
                    Instruction::Call(_, _, to) => {
                    variable_livness.insert_variable_start(
                        *to,
                        AbstractAddress {
                            block_id: *block_id,
                            inststruction: instruction_counter,
                        },
                    )?
                    }
                    _ => (),
                }

                Ok(if instruction_counter == 0 {
                    block.instructions.len()
                } else {
                    instruction_counter - 1
                })
            },
        )?;

        liveness_for_fn.insert(function_id.clone(), livenss);
    }

    Ok(liveness_for_fn)
}

#[cfg(test)]
mod test {
    use crate::cli::load_program;

    use super::*;
    use crate::ir::IrGenerator;
    use anyhow::Result;
    use rstest::rstest;
    use std::path::PathBuf;

    #[rstest]
    fn test_liveness(#[files("./ir_test_programs/test_*.ts")] path: PathBuf) -> Result<()> {
        use crate::{ast::{parser::parse, identifiers::ScopeID, scopes::build_program_scopes}, types::resolve_types, compiler::produce_ir};

        let ir_program = produce_ir(path.to_str().unwrap())?;
        let analysis_result = calculate_livenss(&ir_program);

        insta::assert_snapshot!(
            format!(
                "test_liveness_{}",
                path.file_name().unwrap().to_str().unwrap()
            ),
            format!("{:#?}", analysis_result)
        );


        insta::assert_snapshot!(
            format!(
                "test_liveness_{}",
                path.file_name().unwrap().to_str().unwrap()
            ),
            format!("{:#?}", analysis_result)
        );

        Ok(())
    }
}
