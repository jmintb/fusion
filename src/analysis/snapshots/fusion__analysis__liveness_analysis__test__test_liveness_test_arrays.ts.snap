---
source: src/analysis/liveness_analysis.rs
expression: "format!(\"{:#?}\", analysis_result)"
---
Ok(
    {
        FunctionDeclarationID(
            17,
        ): VariableLiveness {
            variables: {
                SSAID(
                    2,
                ): AbstractAddressRange {
                    start_address: AbstractAddress {
                        block_id: BlockId(
                            4,
                        ),
                        inststruction: 15,
                    },
                    end_addresses: [
                        AbstractAddress {
                            block_id: BlockId(
                                4,
                            ),
                            inststruction: 15,
                        },
                    ],
                },
                SSAID(
                    3,
                ): AbstractAddressRange {
                    start_address: AbstractAddress {
                        block_id: BlockId(
                            4,
                        ),
                        inststruction: 16,
                    },
                    end_addresses: [
                        AbstractAddress {
                            block_id: BlockId(
                                4,
                            ),
                            inststruction: 16,
                        },
                    ],
                },
                SSAID(
                    6,
                ): AbstractAddressRange {
                    start_address: AbstractAddress {
                        block_id: BlockId(
                            4,
                        ),
                        inststruction: 0,
                    },
                    end_addresses: [
                        AbstractAddress {
                            block_id: BlockId(
                                4,
                            ),
                            inststruction: 5,
                        },
                    ],
                },
                SSAID(
                    8,
                ): AbstractAddressRange {
                    start_address: AbstractAddress {
                        block_id: BlockId(
                            4,
                        ),
                        inststruction: 2,
                    },
                    end_addresses: [
                        AbstractAddress {
                            block_id: BlockId(
                                4,
                            ),
                            inststruction: 6,
                        },
                    ],
                },
                SSAID(
                    9,
                ): AbstractAddressRange {
                    start_address: AbstractAddress {
                        block_id: BlockId(
                            4,
                        ),
                        inststruction: 4,
                    },
                    end_addresses: [
                        AbstractAddress {
                            block_id: BlockId(
                                4,
                            ),
                            inststruction: 8,
                        },
                    ],
                },
                SSAID(
                    10,
                ): AbstractAddressRange {
                    start_address: AbstractAddress {
                        block_id: BlockId(
                            4,
                        ),
                        inststruction: 7,
                    },
                    end_addresses: [
                        AbstractAddress {
                            block_id: BlockId(
                                4,
                            ),
                            inststruction: 18,
                        },
                    ],
                },
                SSAID(
                    12,
                ): AbstractAddressRange {
                    start_address: AbstractAddress {
                        block_id: BlockId(
                            4,
                        ),
                        inststruction: 11,
                    },
                    end_addresses: [
                        AbstractAddress {
                            block_id: BlockId(
                                4,
                            ),
                            inststruction: 17,
                        },
                    ],
                },
                SSAID(
                    13,
                ): AbstractAddressRange {
                    start_address: AbstractAddress {
                        block_id: BlockId(
                            4,
                        ),
                        inststruction: 14,
                    },
                    end_addresses: [
                        AbstractAddress {
                            block_id: BlockId(
                                4,
                            ),
                            inststruction: 14,
                        },
                    ],
                },
            },
            variable_moved: {
                SSAID(
                    9,
                ): {
                    BlockId(
                        4,
                    ),
                },
            },
            loans: {},
        },
        FunctionDeclarationID(
            31,
        ): VariableLiveness {
            variables: {
                SSAID(
                    17,
                ): AbstractAddressRange {
                    start_address: AbstractAddress {
                        block_id: BlockId(
                            15,
                        ),
                        inststruction: 0,
                    },
                    end_addresses: [
                        AbstractAddress {
                            block_id: BlockId(
                                15,
                            ),
                            inststruction: 2,
                        },
                    ],
                },
                SSAID(
                    18,
                ): AbstractAddressRange {
                    start_address: AbstractAddress {
                        block_id: BlockId(
                            15,
                        ),
                        inststruction: 1,
                    },
                    end_addresses: [
                        AbstractAddress {
                            block_id: BlockId(
                                15,
                            ),
                            inststruction: 1,
                        },
                    ],
                },
                SSAID(
                    19,
                ): AbstractAddressRange {
                    start_address: AbstractAddress {
                        block_id: BlockId(
                            15,
                        ),
                        inststruction: 11,
                    },
                    end_addresses: [
                        AbstractAddress {
                            block_id: BlockId(
                                15,
                            ),
                            inststruction: 11,
                        },
                    ],
                },
                SSAID(
                    21,
                ): AbstractAddressRange {
                    start_address: AbstractAddress {
                        block_id: BlockId(
                            15,
                        ),
                        inststruction: 3,
                    },
                    end_addresses: [
                        AbstractAddress {
                            block_id: BlockId(
                                15,
                            ),
                            inststruction: 6,
                        },
                    ],
                },
                SSAID(
                    23,
                ): AbstractAddressRange {
                    start_address: AbstractAddress {
                        block_id: BlockId(
                            15,
                        ),
                        inststruction: 4,
                    },
                    end_addresses: [
                        AbstractAddress {
                            block_id: BlockId(
                                15,
                            ),
                            inststruction: 7,
                        },
                    ],
                },
                SSAID(
                    25,
                ): AbstractAddressRange {
                    start_address: AbstractAddress {
                        block_id: BlockId(
                            15,
                        ),
                        inststruction: 5,
                    },
                    end_addresses: [
                        AbstractAddress {
                            block_id: BlockId(
                                15,
                            ),
                            inststruction: 8,
                        },
                    ],
                },
                SSAID(
                    26,
                ): AbstractAddressRange {
                    start_address: AbstractAddress {
                        block_id: BlockId(
                            15,
                        ),
                        inststruction: 10,
                    },
                    end_addresses: [
                        AbstractAddress {
                            block_id: BlockId(
                                15,
                            ),
                            inststruction: 14,
                        },
                    ],
                },
                SSAID(
                    28,
                ): AbstractAddressRange {
                    start_address: AbstractAddress {
                        block_id: BlockId(
                            15,
                        ),
                        inststruction: 12,
                    },
                    end_addresses: [
                        AbstractAddress {
                            block_id: BlockId(
                                15,
                            ),
                            inststruction: 13,
                        },
                    ],
                },
                SSAID(
                    29,
                ): AbstractAddressRange {
                    start_address: AbstractAddress {
                        block_id: BlockId(
                            15,
                        ),
                        inststruction: 20,
                    },
                    end_addresses: [
                        AbstractAddress {
                            block_id: BlockId(
                                15,
                            ),
                            inststruction: 20,
                        },
                    ],
                },
                SSAID(
                    31,
                ): AbstractAddressRange {
                    start_address: AbstractAddress {
                        block_id: BlockId(
                            15,
                        ),
                        inststruction: 17,
                    },
                    end_addresses: [
                        AbstractAddress {
                            block_id: BlockId(
                                15,
                            ),
                            inststruction: 21,
                        },
                    ],
                },
                SSAID(
                    32,
                ): AbstractAddressRange {
                    start_address: AbstractAddress {
                        block_id: BlockId(
                            15,
                        ),
                        inststruction: 19,
                    },
                    end_addresses: [
                        AbstractAddress {
                            block_id: BlockId(
                                15,
                            ),
                            inststruction: 19,
                        },
                    ],
                },
            },
            variable_moved: {
                SSAID(
                    17,
                ): {
                    BlockId(
                        15,
                    ),
                },
                SSAID(
                    19,
                ): {
                    BlockId(
                        15,
                    ),
                },
                SSAID(
                    21,
                ): {
                    BlockId(
                        15,
                    ),
                },
                SSAID(
                    23,
                ): {
                    BlockId(
                        15,
                    ),
                },
                SSAID(
                    25,
                ): {
                    BlockId(
                        15,
                    ),
                },
                SSAID(
                    26,
                ): {
                    BlockId(
                        15,
                    ),
                },
                SSAID(
                    28,
                ): {
                    BlockId(
                        15,
                    ),
                },
            },
            loans: {},
        },
    },
)
